src/user-safe/UserSafe.sol

```diff
@@ -0,0 +1,678 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.24;
+
+import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
+import {SafeERC20, IERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import {IERC20Permit} from "@openzeppelin/contracts/token/ERC20/extensions/IERC20Permit.sol";
+import {ICashDataProvider} from "../interfaces/ICashDataProvider.sol";
+import {SignatureUtils} from "../libraries/SignatureUtils.sol";
+import {ISwapper} from "../interfaces/ISwapper.sol";
+import {IPriceProvider} from "../interfaces/IPriceProvider.sol";
+import {OwnableUpgradeable} from "openzeppelin-contracts-upgradeable/contracts/access/OwnableUpgradeable.sol";
+import {Initializable} from "openzeppelin-contracts-upgradeable/contracts/proxy/utils/Initializable.sol";
+
+/**
+ * @title UserSafe
+ * @author ether.fi [shivam@ether.fi]
+ * @notice User safe account for interactions with the EtherFi Cash contracts
+ */
+contract UserSafe is Initializable, OwnableUpgradeable {
+    using SafeERC20 for IERC20;
+    using SignatureUtils for bytes32;
+
+    enum SpendingLimitTypes {
+        Daily,
+        Weekly,
+        Monthly,
+        Yearly
+    }
+
+    struct WithdrawalRequest {
+        address[] tokens;
+        address recipient;
+        uint96 finalizeTime;
+    }
+
+    struct WithdrawalData {
+        address[] tokens;
+        uint256[] amounts;
+        address recipient;
+        uint96 finalizeTime;
+    }
+
+    struct SpendingLimitData {
+        SpendingLimitTypes spendingLimitType;
+        uint64 renewalTimestamp;
+        uint256 spendingLimit; // in USD with 6 decimals
+        uint256 usedUpAmount; // in USD with 6 decimals
+    }
+
+    bytes32 public constant REQUEST_WITHDRAWAL_METHOD =
+        keccak256("requestWithdrawal");
+    bytes32 public constant APPROVE_METHOD = keccak256("approve");
+    bytes32 public constant SET_SPENDING_LIMIT_METHOD =
+        keccak256("setSpendingLimit");
+    bytes32 public constant SET_INCOMING_SPENDING_LIMIT_METHOD =
+        keccak256("setIncomingSpendingLimit");
+
+    // Address of the USDC token
+    address public immutable usdc;
+    // Address of the weETH token
+    address public immutable weETH;
+    // Address of the Cash Data Provider
+    ICashDataProvider public immutable cashDataProvider;
+    // Address of the price provider
+    IPriceProvider public immutable priceProvider;
+    // Address of the swapper
+    ISwapper public immutable _swapper;
```

Comment:
these are all going to be repeatedly stored in each user safe contracts, which will become an issue if we later want to update them.  any suggestion?

an option is to keep all of these parameters somewhere and make UserSafe contract use them after it pulls them from there

Reply:
I mean we can do that. I can add these parameters in the CashDataProvider and pull from there. But this won't impact much since they are immutable. In future cases as well, if we want to update something, we can deprecate the older immutable variable and create a new one in the V2 contract and deploy. But yes, definitely can move usdc, weETH, priceProvider and swapper into cashDataProvider.

Reply:
Is there any reason for making it immutable?

I am voting for putting these into one place such as `cashDataProvider`

Reply:
I want to keep as many immutables as possible because it is upgradeable and I would want to control setting them while I deploy instead of in the initialize function.

Reply:
Not opposed to it. Let's hear @solipsis 's opinion :)

Reply:
Tell me if I have a misunderstanding here. I thought `immutable` variables where initialized in the constructor. But this will deployed as a beacon proxy so the individual safes will be unable to run that constructor and an initialization style function will be needed?

Reply:
These variables are initialized when we deploy the Beacon (which is the implementation). So yes, we set this in the constructor of the impl contract.

Reply:
Discussed in standup. Immutable variables are initialized in the constructor of the implementation and will work fine. I have no issues here

Reply:
Yes, an immutable variable is embedded directly into the contract’s bytecode.